File: subarray-sum-equals-k/shaun-hua-prefixsum.cpp
/* Huahua idea with prefix sum and hash table
similar to laioffer idea, laioffer has better explanation
cutting prefixsum and hashtable
Running Prefix sum

 Runtime: 52 ms, faster than 63.01% of C++ online submissions for Subarray Sum Equals K.
Memory Usage: 18.1 MB, less than 100.00% of C++ online submissions for Subarray Sum Equals K.

shaun modified with prefixSum - k in mSum hashtable check
runtime 28ms, faster than 98%, mem less than 38%


*/
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        if (nums.empty()) return 0;
        unordered_map<int, int> mSum{{0,1}}; //hashtable: prefixsum to its count
        int prefixSum = 0;
        int count = 0;
        for (int i = 0; i < nums.size(); ++i) {
            prefixSum += nums[i];
            
            // k= prefixSum ending at i - prefixSum ending at j < i?
            //prefixSum_j = prefixSum - k
            // if prefixSum_j in mSum, then found mSum[] candidates
            //  not in, mSum[] will be 0
            //count += mSum[prefixSum - k];
            if (mSum.find(prefixSum -k ) != mSum.end()) {
                count += mSum[prefixSum - k];
            }
            ++mSum[prefixSum];
        }
        return count;
    }
};

File: subarray-sum-equals-k/shaun-hua-prefixsum.java
/*
check on the fly
Runtime: 21 ms, faster than 35.46% of Java online submissions for Subarray Sum Equals K.
Memory Usage: 52.4 MB, less than 5.43% of Java online submissions for Subarray Sum Equals K.

*/
class Solution {
    public int subarraySum(int[] nums, int k) {
        if (nums.length == 0) return 0;
        Map<Integer, Integer> mSum= new HashMap<>(); //hashtable: prefixsum to its count
        mSum.put(0,1);
        int prefixSum = 0;
        int res = 0;
        for (int i = 0; i < nums.length; ++i) {
            prefixSum += nums[i];
            
        

            if (mSum.containsKey(prefixSum -k )) {
                res += mSum.get(prefixSum - k);
            }
            
            int temp = mSum.getOrDefault(prefixSum,0);
            mSum.put(prefixSum, temp+1);
            
        }
        return res;
    }
}
